### PR TITLE
Improve keyboard support for the language switcher

### PR DESCRIPTION
**Pull request by yuliyan**
#### Changes proposed in this Pull Request

This PR improves keyboard support within the language switcher modal.

The main improvements are:

-   Expand search field when start typing
-   Select language if its slug matches search   
-   Confirm selection with Enter key
-   Navigating through search results with keyboard arrow keys

#### Screenshots

Before: 
Screen Capture on 2020-01-06 at 14-24-36

After:
Screen Capture on 2020-01-10 at 12-16-23

#### Testing instructions

-   Check if tests pass
-   Open http://calypso.localhost:3000/
-   Go to Profile → Account Settings → Interface Language
-   Confirm that UI behaves as described

#### TODO

- [x] Expand search field when start typing
- [x] Select language if its slug matches search
- [x] Confirm selection with Enter key
- [x] Navigating through search results with keyboard arrow keys
- [x] Add tests

### CODE REVIEW

**Comment by lsl** [comment]
Test comment: First typed character is lost if you've already selected something with the mouse.

**Comment by yuliyan** [comment]
> Test comment: First typed character is lost if you've already selected something with the mouse.

Fixed in https://github.com/Automattic/wp-calypso/pull/38657/commits/67e1c9e5706873b49b8b456a81bd5c6a3524c31c

File: client/components/language-picker/modal.jsx [hunk 1]
```diff
@@ -185,8 +203,149 @@ export class LanguagePickerModal extends PureComponent {
 		return suggestedLanguages;
 	}
 
+	getLanguagesListColumnsCount() {
+		const wrapper = this.languagesList.current;
+		const wrapperWidth = wrapper.getBoundingClientRect().width;
+		const wrapperChildWidth =
+			wrapper.children.length > 0 ? wrapper.children[ 0 ].getBoundingClientRect().width : 0;
+
+		if ( wrapperWidth === 0 ) {
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply client/components/language-picker/modal.jsx]: You're correct - fixed in 65021a731d202eebca9b20320e0354003ad9f050

File: client/components/language-picker/modal.jsx [hunk 2]
```diff
@@ -156,13 +172,15 @@ export class LanguagePickerModal extends PureComponent {
...
 	getSuggestedLanguages() {
+		// eslint-disable-next-line no-undef
 		if ( ! ( typeof navigator === 'object' && 'languages' in navigator ) ) {
 			return null;
 		}
 
 		const { languages, currentUserLocale } = this.props;
 		const suggestedLanguages = [];
 
+		// eslint-disable-next-line no-undef
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply client/components/language-picker/modal.jsx]: Fixed in 81c06e58ec53d7dd31dd25e81b221904f29a256a

File: client/components/language-picker/modal.jsx [hunk 3]
```diff
@@ -185,8 +203,149 @@ export class LanguagePickerModal extends PureComponent {
...
+	selectLanguageFromSearch( search ) {
+		const filteredLanguages = this.getFilteredLanguages();
+		const exactMatch = filteredLanguages.find( ( { langSlug } ) => langSlug === search );
+
+		if ( exactMatch ) {
+			this.setState( { selectedLanguageSlug: exactMatch.langSlug } );
+			return;
+		}
+
+		const closeMatch = filteredLanguages.reduce( ( currentCloseMatch, language ) => {
```
[the remark this answers is not in the document]
Comment (yuliyan) [reply client/components/language-picker/modal.jsx]: There's additional logic to determine which of the matching languages should be selected - https://github.com/Automattic/wp-calypso/blob/update/improve-language-switcher-keyboard-support/client/components/language-picker/modal.jsx#L246-L248. Basically, my idea was that users will usually search for a language by typing the letters from the beginning of its name, therefore earlier string matches (match index is closer/equal to zero) are prioritized.

[the remark this answers is not in the document]
Comment (akirk) [reply client/components/language-picker/modal.jsx]: I totally agree with your idea but I see a different behavior. Start typing `a`, `l`, `e`. I see Singhalese highighted, not Alemannisch.

**Pull request merged**